server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1060,4 +1154,24 @@ public boolean contains(RelationName tableName) {
         }
         return false;
     }
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public <T extends RelationMetadata> T getRelation(RelationName relation) {
+        SchemaMetadata schemaMetadata = schemas.get(relation.schema());
+        if (schemaMetadata == null) {
+            return null;
+        }
+        RelationMetadata relationMetadata = schemaMetadata.get(relation);
+        if (relationMetadata == null) {
+            return null;
+        }
+        try {
+            return (T) relationMetadata;
+        } catch (ClassCastException e) {
+            throw new OperationOnInaccessibleRelationException(
+                relation,
+                "The relation " + relation.sqlFqn() + " doesn't support the operation");
```

Comment:
Isn't this misleading as it doesn't fail because of a certain "operation"?

Reply:
Not sure what else to call it.

The idea is that you can do callsite casts without type check warning in areas where you know you ought to have a certain `RelationMetadata` type. E.g. resize will only work on shareded tables, so the cluster state update will be able to do `RelationMetadata.Table table = metadata.getRelation(name)` 

If `name` happened to be a view, or blob table it would raise this error.

Reply:
I see, but I think worth adding (debug) log with the cast error?

Reply:
I'd rather leave it out for now and add or adapt if it turns out to be a problem.

We use the same pattern in `schemas.getTableInfo` and I'm not aware that it caused any confusion. The relation name and stacktrace should give enough information to figure out what went wrong.